field/src/extension/binomial_extension.rs

```diff
@@ -0,0 +1,368 @@
+    }
+
+    fn from_canonical_u32(n: u32) -> Self {
+        F::from_canonical_u32(n).into()
+    }
+
+    /// Convert from `u64`. Undefined behavior if the input is outside the canonical range.
+    fn from_canonical_u64(n: u64) -> Self {
+        F::from_canonical_u64(n).into()
+    }
+
+    /// Convert from `usize`. Undefined behavior if the input is outside the canonical range.
+    fn from_canonical_usize(n: usize) -> Self {
+        F::from_canonical_usize(n).into()
+    }
+
+    fn from_wrapped_u32(n: u32) -> Self {
+        F::from_wrapped_u32(n).into()
+    }
+
+    fn from_wrapped_u64(n: u64) -> Self {
+        F::from_wrapped_u64(n).into()
+    }
+
+    fn generator() -> Self {
+        Self(F::ext_multiplicative_group_generator())
+    }
+
+    #[inline(always)]
+    fn square(&self) -> Self {
+        // Specialising mul reduces the computation of c1 from 2 muls
+        // and one add to one mul and a shift
+        *self * *self
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Field for BinomialExtensionField<F, D> {
+    type Packing = Self;
+    // Algorithm 11.3.4 in Handbook of Elliptic and Hyperelliptic Curve Cryptography.
+    fn try_inverse(&self) -> Option<Self> {
+        if self.is_zero() {
+            return None;
+        }
+
+        // Writing 'a' for self, we need to compute a^(r-1):
+        // r = n^D-1/n-1 = n^(D-1)+n^(D-2)+...+n
+        let mut f = Self::ONE;
+        for _ in 1..D {
+            f = (f * *self).frobenius();
+        }
+
+        // g = a^r is in the base field, so only compute that
+        // coefficient rather than the full product.
+
+        let Self(a) = *self;
+        let Self(b) = f;
+        let mut g = F::ZERO;
+        for i in 1..D {
+            g += a[i] * b[D - i];
+        }
+        g *= F::W;
+        g += a[0] * b[0];
+        debug_assert_eq!(Self::from(g), *self * f);
+
+        Some(f * g.inverse())
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Display for BinomialExtensionField<F, D> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{} + {}*a", self.0[0], self.0[1])
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Debug for BinomialExtensionField<F, D> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Display::fmt(self, f)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Neg for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn neg(self) -> Self {
+        Self(
+            self.0
+                .iter()
+                .map(|x| x.neg())
+                .collect::<Vec<F>>()
+                .try_into()
+                .unwrap(),
+        )
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Add for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn add(self, rhs: Self) -> Self {
+        let mut res = self.0;
+        for (r, &rhs_val) in res.iter_mut().zip(&rhs.0) {
```

Comment:
Borrows are not needed here:
```suggestion
        for (r, rhs_val) in res.iter_mut().zip(rhs.0) {
```

Reply:
updated, thx.